simplified solution for trigonometric equations.

Shekharrajak opened the pull request:
###### Master Branch :

```
In [ ]: solveset(cos(x) + cos(3*x) + cos(5*x), x, S.Reals)
Out[ ]: 
⎧        π        ⎫   ⎧        3⋅π        ⎫   ⎧        4⋅π        ⎫   ⎧        2⋅π        ⎫   ⎧        5⋅π        ⎫   ⎧        π        ⎫   ⎧       
⎨2⋅n⋅π + ─ | n ∊ ℤ⎬ ∪ ⎨2⋅n⋅π + ─── | n ∊ ℤ⎬ ∪ ⎨2⋅n⋅π + ─── | n ∊ ℤ⎬ ∪ ⎨2⋅n⋅π + ─── | n ∊ ℤ⎬ ∪ ⎨2⋅n⋅π + ─── | n ∊ ℤ⎬ ∪ ⎨2⋅n⋅π + ─ | n ∊ ℤ⎬ ∪ ⎨2⋅n⋅π +
⎩        2        ⎭   ⎩         2         ⎭   ⎩         3         ⎭   ⎩         3         ⎭   ⎩         3         ⎭   ⎩        3        ⎭   ⎩       

 7⋅π        ⎫   ⎧        5⋅π        ⎫   ⎧        11⋅π        ⎫   ⎧        π        ⎫
 ─── | n ∊ ℤ⎬ ∪ ⎨2⋅n⋅π + ─── | n ∊ ℤ⎬ ∪ ⎨2⋅n⋅π + ──── | n ∊ ℤ⎬ ∪ ⎨2⋅n⋅π + ─ | n ∊ ℤ⎬
  6         ⎭   ⎩         6         ⎭   ⎩         6          ⎭   ⎩        6        ⎭


In [ ]:  solveset(sin(x),  x, S.Reals)
Out[ ]: {2⋅n⋅π | n ∊ ℤ} ∪ {2⋅n⋅π + π | n ∊ ℤ}

In [ ]: solveset(tan(x) - 1, x, S.Reals)
Out[ ]: 
⎧        5⋅π        ⎫   ⎧        π        ⎫
⎨2⋅n⋅π + ─── | n ∊ ℤ⎬ ∪ ⎨2⋅n⋅π + ─ | n ∊ ℤ⎬
⎩         4         ⎭   ⎩        4        ⎭

In [ ]: solveset(cos(x) , x , S.Reals
Out[ ]: 
⎧        π        ⎫   ⎧        3⋅π        ⎫
⎨2⋅n⋅π + ─ | n ∊ ℤ⎬ ∪ ⎨2⋅n⋅π + ─── | n ∊ ℤ⎬
⎩        2        ⎭   ⎩         2         ⎭

In [ ]: solveset(cos(x) + sin(x),  x, S.Reals)
Out[ ]: 
⎧        3⋅π        ⎫   ⎧        7⋅π        ⎫
⎨2⋅n⋅π + ─── | n ∊ ℤ⎬ ∪ ⎨2⋅n⋅π + ─── | n ∊ ℤ⎬
⎩         4         ⎭   ⎩         4         ⎭

```
###### This branch :

```
In [ ]: solveset(cos(x) + cos(3*x) + cos(5*x), x, S.Reals)
Out[ ]: 
⎧      π        ⎫   ⎧n⋅π   π        ⎫   ⎧n⋅π   π        ⎫
⎨n⋅π + ─ | n ∊ ℤ⎬ ∪ ⎨─── + ─ | n ∊ ℤ⎬ ∪ ⎨─── + ─ | n ∊ ℤ⎬
⎩      2        ⎭   ⎩ 2    6        ⎭   ⎩ 2    3        ⎭



In [ ]: solveset(sin(x),  x, S.Reals)
Out[ ]: {n⋅π | n ∊ ℤ}

In [ ]: solveset(tan(x) - 1,  x, S.Reals)
Out[ ]: 
⎧      π        ⎫
⎨n⋅π + ─ | n ∊ ℤ⎬
⎩      4        ⎭


In [ ]: solveset(cos(x),  x, S.Reals)
Out[ ]: 
⎧      π        ⎫
⎨n⋅π + ─ | n ∊ ℤ⎬
⎩      2        ⎭

In [ ]: solveset(cos(x) + sin(x),  x, S.Reals)
Out[ ]: 
⎧      π        ⎫
⎨n⋅π - ─ | n ∊ ℤ⎬
⎩      4        ⎭


```

Also Fixes #10671, #7914, #9606, #9531, #9824
- [x]  Description `imageset union`

Shekharrajak committed:
_solve_trig and _solve_as_rational edited.

Shekharrajak committed:
doctest and test edited.

Shekharrajak committed:
desc on reduce_imageset method

Shekharrajak committed:
changes in reduce_imageset

Shekharrajak commented:
Ping @aktech @hargup @kshitij10496

hargup commented on sympy/solvers/solveset.py:
```diff
@@ -353,39 +353,151 @@ def _is_function_class_equation(func_class, f, symbol):
+
+
+
+def _reduce_imageset(soln):
```
If it is general function (no specific to solveset) it should go in the sets module. Also please add unit test for this function and add examples for this function in the docstring.

Shekharrajak committed:
reduce_imageset is moved to sets.py

Shekharrajak committed:
tan(x) simplified n*pi

Shekharrajak commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -701,36 +701,34 @@ def test_solve_complex_sqrt():
 
 def test_solveset_complex_tan():
     s = solveset_complex(tan(x).rewrite(exp), x)
```
It is solved by `_solve_as_rational` and returned.

Shekharrajak commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -746,19 +744,6 @@ def test_solve_invalid_sol():
 
 
 @XFAIL
-def test_solve_trig_simplified():
```
These were the expected simplified solutions. I edited the `test_solve_trig`.

Shekharrajak commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -878,7 +863,7 @@ def test_conditionset():
 
     assert solveset(Eq(x**2 + x*sin(x), 1), x, domain=S.Reals) == \
-        ConditionSet(x, Eq(x*(x + sin(x)) - 1, 0), S.Reals)
+        ConditionSet(x, Eq(x**2 + x*sin(x) - 1, 0), S.Reals)
```
Both are same actually.

Shekharrajak commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -909,7 +894,7 @@ def test_improve_coverage():
     y = exp(x+1/x**2)
     solution = solveset(y**2+y, x, S.Reals)
-    unsolved_object = ConditionSet(x, Eq((exp((x**3 + 1)/x**2) + 1)*exp((x**3 + 1)/x**2), 0), S.Reals)
+    unsolved_object = ConditionSet(x, Eq(exp(2/x**2)*exp(2*x) + exp(x**(-2))*exp(x), 0), S.Reals)
```
Both are same.

Shekharrajak commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1086,3 +1071,38 @@ def test_issue_8715():
+    assert solveset(sinh(x), x, S.Reals) == FiniteSet(0)
+
+@SKIP("Comparison Error")
+def test_issue_7914():
```
comparison error. I think in LHS `ImageSet(Lambda(n, 2*n*pi - pi/2), S.Integers)` comes but in RHS `ImageSet(Lambda(n, 2*pi*n - pi/2), S.Integers)`. means `2*pi*n ` and `2*n*pi` is making it false.

Shekharrajak commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1086,3 +1071,39 @@ def test_issue_8715():
+    assert solveset(sin(y), y, Interval(0, pi)) == \
+    Intersection(Interval(0, pi), ImageSet(Lambda(n, n*pi), S.Integers))
+
+
```
If you try these examples in current master branch we don't get these simplified solution.

Shekharrajak committed:
comparison error fixed

Shekharrajak commented on sympy/solvers/solveset.py:
```diff
@@ -353,39 +353,69 @@ def _is_function_class_equation(func_class, f, symbol):
+        # try to solve without converting it into exp form
+        # In this time if ConditionSet is returned then it's
+        # expression will be in Trigonometric Function.
+        return _solve_as_poly(f_orig, symbol, domain)
```
`_solve_as_poly` can solve some kind of trig equations in given `domain`.

Shekharrajak committed:
factor_list is used for more simplified solution

Shekharrajak committed:
doctest edited

Shekharrajak commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1086,3 +1071,66 @@ def test_issue_8715():
+
+
+@XFAIL
+def test_simplifed_trig_solution_2():
```
There may be some cases where we need good way to separate 2-3 imageset to make it simple. Right now for these cases it is returning original solution.

Shekharrajak commented:
A [comparison error](https://github.com/sympy/sympy/pull/11188/files#diff-85baa04bbf4e1dfd9128782738e45424R1111) I didn't understand.  Can we use `symbol('n')` instead of `Dummy('n')` .  Some error comes because of these, when I union 2 solutions.

Shekharrajak commented:
PR https://github.com/sympy/sympy/pull/11201 will remove [extra if statement](https://github.com/sympy/sympy/pull/11188/files#diff-eec0422923e8f100745c015cd8fdd6cfR411)

Shekharrajak committed:
changes for radical trig eq

aktech commented on sympy/sets/sets.py:
```diff
@@ -2114,3 +2114,137 @@ def imageset(*args):
+
+def reduce_imageset(soln):
+    """
+    Try to reduce number of imageset in the args.
```
Use ~79 characters in a line.

Shekharrajak replied:
Yeah I am maintaining.

Shekharrajak committed:
shifted reduce_imageset into _union of ImageSet

Shekharrajak commented on sympy/solvers/solveset.py:
```diff
@@ -353,39 +354,113 @@ def _is_function_class_equation(func_class, f, symbol):
+    # issue #11198
+    if f.is_Mul:
+        ind, f = f.as_independent(symbol)
+    fact_list = factor_list(f)[1]
```
` factor` and `factor_list` is used in correct order to get simple solution: [gist](https://gist.github.com/Shekharrajak/17fdcd2320f572fc9fc8674823137e20)

Shekharrajak commented:
Please have a look in these [failed tests](https://travis-ci.org/sympy/sympy/jobs/136877938#L3779) . It works fine. I hope I have replaced the ImageSet expressions `Dummy n` with one `Dummy`.

Shekharrajak commented:
Locally [this test](https://travis-ci.org/sympy/sympy/jobs/137015494#L3792) passes but travis build failed.

Shekharrajak commented:
Ping @aktech @hargup

Shekharrajak committed:
new testcase added

Shekharrajak committed:
new testcase added

debugger22 commented on sympy/sets/sets.py:
```diff
@@ -165,7 +297,10 @@ def _union(self, other):
-        return None
+        try:
+            return self._union_simplify(other)
+        except:
```
What exception are you expecting here?

Shekharrajak replied:
There may be the case when Imageset expr have more than one dummy
            variable present then there may be error in _union_simplify.
            Also it is for simplifying trigonometric eq. solutions (Union of
            Imageset) so one may try with complicated expr in Imageset
            lambda, In that case just return None.

Actually if I remove try block then also it passes all the present testcase.

asmeurer replied:
Please avoid bare except clauses. You should always catch the exception you are looking for, so that unrelated exceptions can bubble up. A bare `except` will catch everything, even things like `RuntimeError` and `KeyboardInterrupt`.

Shekharrajak replied:
Thanks. I will not repeat this again.

debugger22 commented on sympy/solvers/solveset.py:
```diff
@@ -357,39 +359,113 @@ def _is_function_class_equation(func_class, f, symbol):
+        i = 0
+        for a in trig_eq.args:
+            if isinstance(a, TrigonometricFunction) or \
+                                isinstance(a, HyperbolicFunction):
```
Does this work?
```
if (isinstance(a, TrigonometricFunction) or
    isinstance(a, HyperbolicFunction)):
```

debugger22 replied:
Better one:
```
if isinstance(a, (TrigonometricFunction, HyperbolicFunction)):
```

Shekharrajak replied:
Thanks, I will change in next commit.

debugger22 commented on sympy/solvers/solveset.py:
```diff
@@ -357,39 +359,113 @@ def _is_function_class_equation(func_class, f, symbol):
+        # factor the `exp` expression reduce the number of ImageSet
+        # eg. f = 4*sin(x)**3  + 2*sin(x)**2 - 2*sin(x) - 1
+        # compare soln of solveset(f) and solveset(factor(f.rewrite(exp)))
+        if not len(fact_list) == 1 or good_for_factor(fact[0]):
```
This one is easy to understand:

```
if len(fact_list) != 1 or good_for_factor(fact[0]):
```

Shekharrajak replied:
Thanks. I will change in next commit

Shekharrajak committed:
minor changes

minor changes

minor change2

minor change3

one testcase added

minor changes in docs

checking imageset for _union_simplify

hargup commented on sympy/sets/sets.py:
```diff
@@ -165,7 +307,16 @@ def _union(self, other):
+        try:
+            return self._union_simplify(other)
+        except:
+            # There may be the case when Imageset expr have more than one
```
@Shekharrajak Adding a comment doesn't [fix the issues with bare except clauses](https://github.com/sympy/sympy/pull/11188/files#r68429752). People find the specific error which is raised for the condition you are describing and specifically catch that.

Shekharrajak replied:
Thanks, I will not repeat this again.

hargup commented on sympy/sets/sets.py:
```diff
@@ -153,6 +153,148 @@ def isdisjoint(self, other):
+            if Poly(lamb_expr, n_final).is_linear:
+                final_expr.append(lamb_expr)
+            else:
+                raise NotImplementedError('mplemented for linear Imageset expr.\
```
Typo

Shekharrajak replied:
thanks, I'll fix this in next commit.

Shekharrajak commented:
I don't know why this PR sometimes pass all checks and sometimes not.  Testcases works fine locally.
If someone finds any mistake where any variable is changing, then let me know. I have checked it many times but not found the mistake(There is only one dummy `n` in final soln so I don't think problem is in dummy ). Build is this : https://travis-ci.org/sympy/sympy/builds/140861971 . Travis pass the failed check if you restart the build but problem is why it is failing sometimes.

kshitij10496 commented on sympy/sets/sets.py:
```diff
@@ -153,6 +153,181 @@ def isdisjoint(self, other):
         """
         return self.is_disjoint(other)
 
+    def _union_simplify(self, other):
```
Since this method is not general ( it can simplify only specific types of union), I think all of this code should be moved to `ImageSet` class in sets/fancysets.py .

Shekharrajak replied:
I saw union of imageset is coming here [_union](https://github.com/sympy/sympy/pull/11188/files#diff-7b3e9b0852f1b1d094f87233f56e201fR331). 
Should we have `_union` in `fancysets/ImageSet` like we have [`_intersection`](https://github.com/Shekharrajak/sympy/blob/fae9f28623fb9bd81baef08cc0582aba9034a26f/sympy/sets/fancysets.py#L366) ?

kshitij10496 replied:
Yes !
I think that would be the best place to put this code.

kshitij10496 commented on sympy/sets/sets.py:
```diff
@@ -153,6 +153,181 @@ def isdisjoint(self, other):
 
+    def _union_simplify(self, other):
+        """
+        Try to reduce number of imageset in the args. Helper method for
```
Can you add the definition of this function stating the abstraction it provides ?

kshitij10496 commented on sympy/sets/sets.py:
```diff
@@ -153,6 +153,181 @@ def isdisjoint(self, other):
+        from sympy.core import pi
+        from sympy.polys import Poly
+
+        # don't try to reduce imageset(s) from the same union.
```
About which `same union` are we talking here ?
What does this line mean ?

Shekharrajak replied:
I explained it from next line
```
        # `_solve_trig` function defined in `solveset.py`
        # solves the factors of trig eq and union of these soln is final ans,
        # soln of one factor (say f1) should be clubbed with soln of another
        # factor(say f2) and so on. ImageSets of f1 or f2 should not be
        # reduced(clubbed) with it's own args.
```
My intention is to tell why it returns `self` if  `other` is EmptySet and vice versa.
I will remove first line, it is confusing.
Thanks.

kshitij10496 replied:
> My intention is to tell why it returns self if other is EmptySet and vice versa.

I think this is implicit (according to the definition of union).
I don't think we need to state it explicitly here.
Even if there is an agreement on stating this, I believe this should be stated clearly in the docstrings of the function.

Shekharrajak replied:
Okay I will shift these lines there.

kshitij10496 commented on sympy/sets/sets.py:
```diff
@@ -165,7 +340,37 @@ def _union(self, other):
+                    if not isinstance(self_img, ImageSet):
+                        return False
+                return True
+            else:
```
This `else` clause can be removed.

Shekharrajak replied:
Yes! thanks

kshitij10496 commented on sympy/sets/sets.py:
```diff
@@ -165,7 +340,37 @@ def _union(self, other):
+                return False
+        try:
+            if _check_imageset(self) and _check_imageset(other):
+                return self._union_simplify(other)
```
Why would we want to have a `_union_simplify` method for all our `Set` objects ?
Can you elaborate on the integrity of this behaviour ?

Shekharrajak replied:
[same comment](https://github.com/sympy/sympy/pull/11188/files#r68915598)

kshitij10496 replied:
I am asking a totally different question here.

Shekharrajak replied:
>Why would we want to have a _union_simplify method for all our Set objects ?

`_union_simplify` will be executed when `self ` and `other` is imageset or union of imageset ( `_check_imageset` returns True when self/other contains only imageset(s)) .

kshitij10496 replied:
Okay.
And what is the plausibility of this bound method for a `FiniteSet` object (or any object with `Set` as its base class) ?

Shekharrajak replied:
>And what is the plausibility of this bound method for a FiniteSet object 

For `FiniteSet` union [these lines]
(https://github.com/Shekharrajak/sympy/blob/fae9f28623fb9bd81baef08cc0582aba9034a26f/sympy/sets/sets.py#L1401) is being used.

kshitij10496 commented on sympy/sets/sets.py:
```diff
@@ -165,7 +340,37 @@ def _union(self, other):
+            NotImplementedError,
+            PolynomialError
+        ):
+            # `_union_simplify` is for simplifying trigonometric eq. solution
```
The behaviour of the `_union` method should be independent of the details of how the class was instantiated.
This line contradicts this abstraction.

Shekharrajak replied:
I shifted the code in `fancysets/ImageSet._union` . Now It works for all other cases.

kshitij10496 commented on sympy/solvers/solveset.py:
```diff
@@ -63,18 +64,19 @@ def _invert(f_x, y, x, domain=S.Complexes):
 
-    >>> invert_complex(exp(x), y, x)
-    (x, ImageSet(Lambda(_n, I*(2*_n*pi + arg(y)) + log(Abs(y))), Integers()))
+    >>> pprint(invert_complex(exp(x), y, x), use_unicode = False)
```
Pretty printing is not needed, I suppose.
These changes make the doc strings look irregular (with some results being pretty-printed while others display the returned values from the function calls)
Can you share the particular reason why you have made this change ?

@asmeurer @hargup @aktech :  What's the convention for writing examples in docstrings ?

Shekharrajak replied:
It was failing when I did `bin/doctest` locally(but it works fine). Then I tried `pprint` then it passed.(sometimes fail, sometimes pass) I will restore these changes.
Thanks.

Shekharrajak committed:
minor changes

is_ImageSet using

whitespace removed

Shekharrajak committed:
_union modified

Shekharrajak commented on sympy/sets/sets.py:
```diff
@@ -1202,28 +1202,32 @@ def reduce(args):
         # ===== Pair-wise Rules =====
         # Here we depend on rules built into the constituent sets
-        args = set(args)
+        args = FiniteSet(*args)
```
I want `args` always in same order. May be there is better way to do it.

hargup replied:
`FiniteSet` doesn't guarantee an order. A tuple is already ordered, you can create a copy of `args` and use it later where you need it.

_General Comment_: You should avoid using undocumented behavior  of any function as they might change in future. For example the current algorithm used in converting list to `FiniteSet` might maintain order but a future implementations might change this behavior breaking your code.

/cc @kshitij10496

hargup replied:
Sometimes the undocumented behavior depends on random seed which can lead to crazy bugs which sometimes appear in the code and sometime they don't. Maybe the [problem you mentioned above](https://github.com/sympy/sympy/pull/11188#issuecomment-229264778) arises because of that.

Shekharrajak replied:
Thanks @hargup . I did this changes because of [problem I mentioned](https://github.com/sympy/sympy/pull/11188#issuecomment-229264778) After this changes (using finiteset, instead of list to always get elements in one order) test cases passes(but still I felt there is something that causes the problem). I will try some alternate way.

hargup replied:
I thought of providing an alternate implementation of this part myself but I'm running to strange errors related to types. Will try again tomorrow morning.

Shekharrajak replied:
Thanks @hargup , I am also trying to find out alternate way.

Shekharrajak replied:
I am trying to use only `frozenset` and `ordered` in [line 1205](https://github.com/sympy/sympy/pull/11188/files#diff-7b3e9b0852f1b1d094f87233f56e201fR1205), but still error in coming (each time different assertion error for different python version).

```
args = list(ordered(frozenset(tuple(args)), Set._infimum_key))
# or 
args = set(frozenset(args))
```

Shekharrajak replied:
few months before I came across similar problem ([comparison error](https://github.com/sympy/sympy/pull/10733#issuecomment-216266502)), but that looks different case.

Shekharrajak replied:
Ping @asmeurer , @jksuom , @smichr . Any alternate way of doing [this](https://github.com/sympy/sympy/pull/11188/files#diff-7b3e9b0852f1b1d094f87233f56e201fR1205) ?

Shekharrajak commented on sympy/solvers/solveset.py:
```diff
@@ -357,39 +359,110 @@ def _is_function_class_equation(func_class, f, symbol):
+    fact_list = factor_list(Poly(f))[1]
+
+    soln = S.EmptySet
+    for fact in fact_list:
```
[see this examples](https://gist.github.com/Shekharrajak/17fdcd2320f572fc9fc8674823137e20) to understand how factoring can reduce the no. of imageset (give simplifed soln).

Shekharrajak committed:
new test added

hargup commented on sympy/sets/fancysets.py:
```diff
@@ -468,6 +468,123 @@ def _intersect(self, other):
+                if Poly(lamb_expr, var_self).is_linear:
+                    self_expr = lamb_expr
+                else:
+                    msg = 'Not Implemented for non linear ImageSet. \
```
There error messages won't ever show up because you are catching every `NotImplementedError`, rather you should return `None` here and use `try` on only a small block of code. Error messages and tracebacks are very useful while debugging. Hence as a general practice you should be very precise while catching errors. By precision I mean you should catch **only** the errors you want to catch and nothing else. If you can apply try-catch to 3 lines of code instead of 9, then do it. If you can only catching one error instead of three does your work then do it.

/cc @kshitij10496

hargup commented:
@Shekharrajak Your code gives wrong answer for the following example:
```
In [10]: Union(ImageSet(Lambda(n, 4*n*pi + pi/4), S.Integers) + ImageSet(Lambda(n, 4*n*pi + 5*pi/4), S.Integers))           
Out[10]: ImageSet(Lambda(n, pi*n + pi/4), Integers())
```
You can create little more general union rule as follows. If the image two `ImageSets` are of form `a*n*pi + p`, `a*n*pi + q`, for `n` in Integers and `|p - q|` is `a/2` then union is `(a/2)*n*pi + p`

Shekharrajak commented:
Thanks @hargup , I am trying to improving it.

Shekharrajak committed:
improved imageset union

Shekharrajak commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1087,13 +1106,79 @@ def test_issue_10555():
+    img2 = ImageSet(Lambda(n, n * pi / 2 + pi / 6), S.Integers)
+    img3 = ImageSet(Lambda(n, n * pi / 2 + pi / 3), S.Integers)
+    soln = Union(img1, img2, img3)
+    assert solveset(cos(x) + cos(3 * x) + cos(5 * x), x, S.Reals) == soln
```
One can read this [gist](https://gist.github.com/Shekharrajak/a5efc840d9a7d3062289f2d9c5f20b16) to understand how the `imageset. _union` is helpful to reduce the number of Imagesets.

hargup commented on sympy/solvers/solveset.py:
```diff
@@ -357,39 +359,110 @@ def _is_function_class_equation(func_class, f, symbol):
+    if _is_function_class_equation(TrigonometricFunction, f, symbol):
+        f = trigsimp(f)
+    # trigsimp is not defined for hyperbolic
+    # TODO trighsimp function for Hyperbolic Functions if required
```
typo, `trighsimp` should be `trigsimp`.

Shekharrajak replied:
Thanks @hargup . If we want to simplify hyperbolic then we can use `hyper_as_trig` defined in [simplify/fu.py](https://github.com/sympy/sympy/blob/master/sympy/simplify/fu.py#L2096). I will do required changes.

hargup commented:
@Shekharrajak Does your code handles the corner cases that I mentioned above? If yes please add a tests for those cases.

Shekharrajak commented:
@hargup , That [test case](https://github.com/sympy/sympy/pull/11188#issuecomment-234789616) is [here](https://github.com/sympy/sympy/pull/11188/files#diff-e21906560346d439e577d8c1be6b3eceR1061)

Shekharrajak committed:
simplify the hyperbolic function in _solve_trig

Shekharrajak committed:
hyperbolic testcase added

Shekharrajak committed:
minor change in docs

hargup commented:
@Shekharrajak Please also post the test coverage for your code.

Shekharrajak commented:
@hargup , This is the link for coverage report : https://shekharrajak.github.io/sympy_11188_coverage_report/ .

hargup commented:
> This is the link for coverage report : https://shekharrajak.github.io/sympy_11188_coverage_report/sympy_solvers_tests_test_solveset_py.html

This coverage report is "wrong". The coverage report shows which lines of the code are being run when we run the test suite, you have run the coverage on the test suit itself which is kinda useless. You need to run the following and if most the lines of the code adding in this PR are covered or not.
```
./bin/coverage_report.py sympy/sets/
```

hargup commented:
If you run the coverage report you'll see that many important lines like (530, [584](https://github.com/sympy/sympy/pull/11188/files#diff-c1c17bdd28aafd400bf9415c7d048382R584) and [585](https://github.com/sympy/sympy/pull/11188/files#diff-c1c17bdd28aafd400bf9415c7d048382R585)) in `_union` in `fancysets.py` are not covered.

Shekharrajak commented:
@hargup , are you talking about this : https://shekharrajak.github.io/sympy_11188_coverage_report/

https://shekharrajak.github.io/sympy_11188_coverage_report/sympy_sets_fancysets_py.html

I did `./bin/coverage_report.py sympy/solvers/tests/test_solveset.py`

Shekharrajak committed:
minor change

Shekharrajak commented:
Link for coverage_report : 
solveset : https://shekharrajak.github.io/sympy_11188_coverage_report/sympy_solvers_solveset_py.html#n404

fancyset/imageset: https://shekharrajak.github.io/sympy_11188_coverage_report/sympy_sets_fancysets_py.html#n559

kshitij10496 commented on sympy/sets/fancysets.py:
```diff
@@ -483,6 +483,128 @@ def _intersect(self, other):
+                # Add more ways to club `self` and `other`
+                return ans
+
+            except (
```
This is an interesting design.
You are [raising a `NotImplementedError` exception](https://github.com/sympy/sympy/pull/11188/files#diff-c1c17bdd28aafd400bf9415c7d048382R564) in the `try` block and catching the raised error in the `except` clause.
Ping @hargup @aktech

kshitij10496 replied:
Not addresed

kshitij10496 commented on sympy/sets/fancysets.py:
```diff
@@ -483,6 +483,128 @@ def _intersect(self, other):
+                var_self = var_self[0]
+                var_final = other.lamda.variables
+                if len(var_final) > 1:
+                    # Not Implemented for multiple lambda variables.
```
Can you move both of the `if` clauses into a single check ?

Shekharrajak replied:
Done.

kshitij10496 commented on sympy/sets/fancysets.py:
```diff
@@ -483,6 +483,128 @@ def _intersect(self, other):
+
+                # Extracting expr
+                lamb_expr = self.lamda.expr
+                if Poly(lamb_expr, var_self).is_linear:
```
I think you need to handle an exception here.

`Poly(sin(x), x)` raises `PolynomialError`.

Shekharrajak replied:
Already catching `PolynomialError`, [here](https://github.com/sympy/sympy/pull/11188/files#diff-c1c17bdd28aafd400bf9415c7d048382R591)

kshitij10496 replied:
Can you put the `try-except` block just around this statement ?

Its clearer that way

kshitij10496 commented on sympy/sets/fancysets.py:
```diff
@@ -483,6 +483,128 @@ def _intersect(self, other):
+                else:
+                    # Not Implemented for non linear ImageSet.
+                    return None
+                if base_self.is_superset(base):
```
You are initally checking for base set equality [here](https://github.com/sympy/sympy/pull/11188/files#diff-c1c17bdd28aafd400bf9415c7d048382R563).
Now, you checking for `superset`.

Can you elaborate on the need for this check ?

Shekharrajak replied:
Thanks, I forgot to remove this `if` block. Now it is removed.

kshitij10496 commented on sympy/sets/fancysets.py:
```diff
@@ -483,6 +483,128 @@ def _intersect(self, other):
+                    # take the superset
+                    base = base_self
+
+                lamb_expr = other.lamda.expr
```
I think you are repeating yourself here.

kshitij10496 commented on sympy/sets/fancysets.py:
```diff
@@ -483,6 +483,128 @@ def _intersect(self, other):
+            except TypeError:
+                raise TypeError
+
+        def _attempt_1(self_expr, final_expr, var_self, var_final, base):
```
Rather than passing the variables as arguments, you can use a `Dummy` for both the `expr`s.

Shekharrajak replied:
What is the problem in this? 

>you can use a Dummy for both the expr

I don't know about this, how to do that ?

kshitij10496 replied:
Use `subs`

kshitij10496 commented on sympy/sets/fancysets.py:
```diff
@@ -483,6 +483,128 @@ def _intersect(self, other):
+
+        def _attempt_1(self_expr, final_expr, var_self, var_final, base):
+            var_term = self_expr.coeff(var_self)
+            if var_term == 0:
```
It would make more sense if you can merge this clause with this [check](https://github.com/sympy/sympy/pull/11188/files#diff-c1c17bdd28aafd400bf9415c7d048382R589)

Shekharrajak replied:
Now it is not needed, I have removed it.

kshitij10496 commented on sympy/sets/fancysets.py:
```diff
@@ -483,6 +483,128 @@ def _intersect(self, other):
+                iterable = iter(base)
+                return imgset.lamda(next(iterable))
+            except TypeError:
+                raise TypeError
```
Elaborate here please !

Shekharrajak replied:
Previously I wrote some msg, along with errors. I removed others but forgot to remove this raise. Now I have removed it. Thanks.

kshitij10496 commented:
General comments on the PR:

1. Defining functions within a function (and that too an **internal** method).
2. Lot of the clauses are repeated.
3. The use `try-except` clause can be improved.
4. The examples can be moved to the examples section of `ImageSet` class.

A general query: Should we add docstings for internal functions (such as `_union`) as well ?
Ping @sympy/sympy-solvers

hargup commented on sympy/sets/fancysets.py:
```diff
@@ -483,6 +492,108 @@ def _intersect(self, other):
+            return ans
+
+        if isinstance(other, ImageSet):
+            try:
```
Please use `try`-`except` precisely on the lines you need to use them, otherwise they make it harder to debug the code. (I guess I've mentioned this already somewhere)

hargup commented on sympy/sets/fancysets.py:
```diff
@@ -483,6 +492,108 @@ def _intersect(self, other):
+            try:
+                base = other.base_set
+                if base != self.base_set:
+                    raise NotImplementedError
```
You can directly return `None` here.

hargup commented on sympy/sets/tests/test_sets.py:
```diff
@@ -1041,11 +1041,78 @@ def test_issue_8257():
+    uni = Union(img1, img2, evaluate=False)
+    assert Union(img1, img2) == uni
+
+    img2 = ImageSet(Lambda((n, m), 2 * n * pi + pi), S.Integers)
```
Why is it a imageset of two variables?

Shekharrajak replied:
ImageSet union is not implemented for multiple variable. this it the testcase for that. It will be returned as it is.

hargup commented:
@Shekharrajak You have more wrong results 
```python
In [6]: Union(ImageSet(Lambda(n, 2*n*pi + pi/4), S.Integers) + ImageSet(Lambda(n, 4*n*pi + 5*pi/4), S.Integers))
Out[6]: 
⎧      π        ⎫
⎨π⋅n + ─ | n ∊ ℤ⎬
⎩      4        ⎭

```
Please explain your logic very clearly and specify in which cases it will work and in which won't.

hargup commented on sympy/solvers/solveset.py:
```diff
@@ -357,39 +359,106 @@ def _is_function_class_equation(func_class, f, symbol):
+        # search for exp(I*x) terms
+        flag = False
+        atoms = list(g.atoms(exp))
+        atoms += list(h.atoms(exp))
```
use `atoms.extend(h.atoms(exp))`

hargup commented on sympy/solvers/solveset.py:
```diff
@@ -357,39 +359,106 @@ def _is_function_class_equation(func_class, f, symbol):
+            # make it independent from exp(I*symbol)
+            g, h = g.subs(exp(I*symbol), y), h.subs(exp(I*symbol), y)
+
+            solns = _solveset(g, y, domain) - _solveset(h, y, domain)
```
Why are you using `_solveset` instead of `solveset`?

Shekharrajak replied:
The equations are solvable using `_solveset`. I was suggested to use lower level functions, previously.

hargup commented on sympy/solvers/solveset.py:
```diff
@@ -357,39 +359,106 @@ def _is_function_class_equation(func_class, f, symbol):
+            if isinstance(solns, FiniteSet):
+                result = Union(*[invert_complex(exp(I*symbol), s, symbol)[1]
+                       for s in solns])
+                return result.intersection(domain) \
```
This compound statement doesn't look very readable, better expand the `if else` clause.

hargup commented on sympy/solvers/solveset.py:
```diff
@@ -357,39 +359,106 @@ def _is_function_class_equation(func_class, f, symbol):
+        f = trigsimp(f)
+    # trigsimp is not defined for hyperbolic
+    elif _is_function_class_equation(HyperbolicFunction, f, symbol):
+        try:
```
Can you explain how does this part works?

Shekharrajak replied:
```
In [1]: f = -sinh(x)**2 + cosh(x)**2

In [2]: from sympy.simplify.fu import hyper_as_trig, fu

In [3]: t, f = hyper_as_trig(f)

In [4]: t
Out[4]: 
   2                 2           
sin (Dummy₁₈⋅x) + cos (Dummy₁₈⋅x)

In [5]: f
Out[5]: <function sympy.simplify.fu.<lambda>>

In [7]: fu(t)
Out[7]: 1  # so f in simplified to 1 
```

Mor detail is [here](https://github.com/sympy/sympy/blob/master/sympy/simplify/fu.py#L2112)

hargup commented on sympy/solvers/solveset.py:
```diff
@@ -357,39 +359,106 @@ def _is_function_class_equation(func_class, f, symbol):
+        # In this time if ConditionSet is returned then it's
+        # expression will be in Trigonometric Function(original eq).
+        return _solve_as_poly(f_orig, symbol, domain)
+    return soln.intersection(domain) if domain.is_subset(S.Reals) else soln
```
Same comment as above, use explicit `if`- `else` blocks.

hargup commented on sympy/solvers/solveset.py:
```diff
@@ -357,39 +359,106 @@ def _is_function_class_equation(func_class, f, symbol):
-    g, h = g.subs(exp(I*symbol), y), h.subs(exp(I*symbol), y)
-    if g.has(symbol) or h.has(symbol):
-        return ConditionSet(symbol, Eq(f, 0), S.Reals)
+    from sympy import factor, factor_list
```
Using `factor_list` should be used generally, instead of only for trig functions. Please make changes in `_solveset` where we take union of solutions of `g` and `h` for equations of form `g*h`. [This](https://github.com/sympy/sympy/pull/11188/files#diff-eec0422923e8f100745c015cd8fdd6cfR711) is probably a good place.

I wonder why don't we use `factor_list` already, I guess it is because I though `together` is a good as `factor_list`. Thanks for digging this out.

Shekharrajak replied:
I will do that in another PR, because in few cases it doesn't work properly(e.g. const factors are not `number`).

hargup replied:
Open an issue about it so that we remember.

hargup replied:
Also mention the issues you are facing while implementing that.

Shekharrajak committed:
improved imageset union and more testcases added

kshitij10496 commented on sympy/sets/fancysets.py:
```diff
@@ -483,6 +492,127 @@ def _intersect(self, other):
+        self._union(other) returns a new, joined ImageSet if self knows how
+        to join itself with other, otherwise it returns ``None``.
+
+        Examples
```
These examples are already present in the docstrings of `ImageSet`.
Hence, they are redundant.
Moreover, is adding docstring to internal functions a good practice ?
Ping @hargup @aktech

hargup replied:
Using docstrings is a good practice, I don't see how using them for internal functions should be a problem, also redundancy is not an issue as long as it is adding clarity to the explanation.

Shekharrajak commented:
@hargup , these are the base cases : 

```
# img1 = ImageSet(Lambda(n, a*n + b), S.Integers)
# img2 = ImageSet(Lambda(n, c*n + d), S.Integers)
In [1]: img1 = ImageSet(Lambda(n, 4*n + 4), S.Integers)

In [2]: img2 = ImageSet(Lambda(n, 4*n), S.Integers)
# when a == c and (b - d == a) then ans is img2.
In [3]: Union(img1, img2)
Out[3]: {4⋅n | n ∊ ℤ}

In [4]: img1 = ImageSet(Lambda(n, 4*n + 2), S.Integers)
# when a == c and (b - d) == a/2, means value is  a/2 * n 
In [5]: Union(img1, img2)
Out[5]: {2⋅n | n ∊ ℤ}

In [6]: img1 = ImageSet(Lambda(n, n + 2), S.Integers)
# when img1 contains all the value of img2. When a != c
# and (c + (b - d)) mod a == 0, means (c + (b-d))*n may be multiple of a*n
#  after checking it for 3 values, returns img1.
In [7]: Union(img1, img2)
Out[7]: {n + 2 | n ∊ ℤ}

In [8]: img1 = ImageSet(Lambda(n, 2*n*pi), S.Integers)

In [11]: img2 = ImageSet(Lambda(n, 2*n*pi + pi), S.Integers)

In [12]: Union(img1, img2)
Out[12]: {π⋅n | n ∊ ℤ}

# For all others return as it is (without evaluating). 
```

hargup commented:
I don't quite understand the third case that you have mentioned, it says:
```
# when img1 contains all the value of img2. When a != c
# and (c + (b - d)) mod a == 0, means (c + (b-d))*n may be multiple of a*n
#  after checking it for 3 values, returns img1.
```
when img1 contains all the value of img2, then solution should be `img1` isn't it? Also how does `(c + (b-d))` being multiple of `a` determine that `img1` contains all the values of `img2`.

Shekharrajak commented:
@hargup ,  Actually I wanted to say this 
(Sorry for the line `when img1 contains all the value of img2` )

```
# img1 = ImageSet(Lambda(n, a*n + b), S.Integers)
# img2 = ImageSet(Lambda(n, c*n + d), S.Integers)
#  When a != c  and (c + (b - d)) mod a == 0, means (c + (b-d))*n may be multiple of a*n. 
#  After checking it for 3 values, returns img1.

In [1]: Union(ImageSet(Lambda(n, 3*n), S.Integers), ImageSet(Lambda(n, 6*n), S.Integers))
Out[1]: {3⋅n | n ∊ ℤ}

In [2]: Union(ImageSet(Lambda(n, 3*n ), S.Integers), ImageSet(Lambda(n, 7*n ), S.Integers))
Out[2]: {3⋅n | n ∊ ℤ} ∪ {7⋅n | n ∊ ℤ}

In [3]: Union(ImageSet(Lambda(n, 3*n ), S.Integers), ImageSet(Lambda(n, n+2 ), S.Integers))
Out[3]: {n + 2 | n ∊ ℤ}
# -------------------------------------------------------------
# 5*n + 5/4 ==> 5*n + 1 + 1/4 and (5 + 1 )%1 ==0(here a=1 , b=5, abs(b-d) = 1).
# 5*n + 1 + 1/4 is in n + 1/4 
# cross check using img1.superset(img2) == true
img1 = ImageSet(Lambda(n, n + S(1)/4 ), S.Integers)
img2 = ImageSet(Lambda(n, 5*n + S(5)/4 ), S.Integers)

In [4]: Union(img1, img2)
Out[4]: {n + 1/4 | n ∊ ℤ}

# -------------------------------------------------------------
# 5*n + 5/4 ==> 5*n + 1 + 1/4 and (5 + 1) % 2  == 0 (here a=2 , b=5, abs(b-d) = 1)
# but for n=2 it is 10 + 5/4 => 11 + 1/4 which can't be generated using 2*n + 1/4.
#  (This is the reason why I am checking for n = 1, 2, 3 
# means 5*1 + 5/4, 5*2 + 5/4, 5*3 + 5/4 is in 2*n + 1/4 or not)
# cross check:
# img1.issuperset(img2) is false
img1 = ImageSet(Lambda(n, 2*n + S(1)/4 ), S.Integers)
img2 = ImageSet(Lambda(n, 5*n +S(5)/4), S.Integers)
In [5]: Union(img1, img2)
Out[5]: {2⋅n + 1/4 | n ∊ ℤ} ∪ {5⋅n + 5/4 | n ∊ ℤ}

# Here we should also check that if b==0 then d % a should be 0.
# Union(ImageSet(Lambda(n, 3*n ), S.Integers), ImageSet(Lambda(n, 6*n+2 ), S.Integers)) 
# should be returned as it is.

```

Let me know if you have good logic for this.

I think it is better to use `issuperset` (for this case i.e. a != c case) and return img1. Should I change it ?

Shekharrajak committed:
using is_superset when coeff of lambda variable is not same

more testcases

Shekharrajak commented:
@hargup , now it looks good to go.

hargup commented:
In your own example:
```
In [3]: Union(ImageSet(Lambda(n, 3*n ), S.Integers), ImageSet(Lambda(n, n+2 ), S.Integers))
Out[3]: {n + 2 | n ∊ ℤ}
```
the solution is plain wrong, since `4` is in `ImageSet(Lambda(n, n+2 ), S.Integers)` but not in `ImageSet(Lambda(n, 3*n ), S.Integers)`. 

Also, I don't understand the following logic you have given. Can you give a proof of it and illustrate why it should be true.

> img1 = ImageSet(Lambda(n, a*n + b), S.Integers)
> img2 = ImageSet(Lambda(n, c*n + d), S.Integers)
>  When a != c  and (c + (b - d)) mod a == 0, means (c + (b-d))*n may be multiple of a*n. 
>  After checking it for 3 values, returns img1.

Shekharrajak commented:
> the solution is plain wrong, since 4 is in ImageSet(Lambda(n, n+2 ), S.Integers) but not in ImageSet(Lambda(n, 3*n ), S.Integers).

```
img1 = ImageSet(Lambda(n, n+2 ), S.Integers) 
img2 = ImageSet(Lambda(n, 3*n ), S.Integers)
```

[union](https://en.wikipedia.org/wiki/Union_(set_theory)) is img1, since img1 contains `x` where `x ∊ img1 or x ∊ img2` .

> Also, I don't understand the following logic you have given. Can you give a proof of it and illustrate why it should be true.

I have removed this logic in [this commit](https://github.com/sympy/sympy/pull/11188/commits/4fbc023b1fb7d336d5e4bdecbaca1d608171603b#diff-c1c17bdd28aafd400bf9415c7d048382R554) . Now I am using `issuperset`, which does the same thing in better way.

In above case 

```
In [6]: img1 = ImageSet(Lambda(n, n+2 ), S.Integers) 

In [7]: img2 = ImageSet(Lambda(n, 3*n ), S.Integers)

In [8]: img1.issuperset(img2)
Out[8]: True
```

We can get all the values of `img2` using `img1` so union of these is `img1`.

Thanks for review.  I am fixing conflict.

Shekharrajak committed:
few xfail

hargup commented:
I'm sorry regarding the union thing, the result it correct.

Shekharrajak commented:
No problem. Please restart this build https://travis-ci.org/sympy/sympy/builds/152597422

Shekharrajak commented on sympy/sets/sets.py:
```diff
@@ -1202,28 +1202,32 @@ def reduce(args):
                             new_set = set((new_set, ))
-                        new_args = (args - set((s, t))).union(new_set)
+                        remv = set((s, t))
+                        new_args = (set(args) - remv).union(new_set)
```
If we get new set then inner for loop L1212 `break`.

Shekharrajak commented on sympy/sets/sets.py:
```diff
@@ -1202,28 +1202,32 @@ def reduce(args):
                 if new_args:
+                    new_args = FiniteSet(*[ar for ar in new_args])
                     args = new_args
                     break
```
After the [break ](https://github.com/sympy/sympy/pull/11188/files#diff-7b3e9b0852f1b1d094f87233f56e201fR1220)
it comes here and uses new `args`, `break` this `L1208` for loop and starts again.

Shekharrajak replied:
@hargup , after the modification of `args` set , `for` loop will be started again. Is it good for now ?

hargup commented on sympy/sets/sets.py:
```diff
@@ -1202,28 +1202,32 @@ def reduce(args):
                         break
                 if new_args:
+                    new_args = FiniteSet(*[ar for ar in new_args])
                     args = new_args
```
@Shekharrajak as I said above the actual issue is here, modifying the list your are iterating. You need to rewrite the whole block so that this doesn't happen.

Shekharrajak replied:
There is [`break`](https://github.com/sympy/sympy/pull/11188/files#diff-7b3e9b0852f1b1d094f87233f56e201fR1225)  after `args = new_args` . So it comes out of the   [Line 1208 `for` loop](https://github.com/sympy/sympy/pull/11188/files#diff-7b3e9b0852f1b1d094f87233f56e201fR1208) and since `while(new_args)`  is not none,   `for` loop [L1208] will start again with modified `args`.  

Whenever `new_set = s._union(t)` new_set is not none (means reduced form comes) , it will modify the `new_args` and `break` the loop, to start again fresh .

`while(new_args)` --  new_args is for just ([ `false`](https://github.com/sympy/sympy/pull/11188/files#diff-7b3e9b0852f1b1d094f87233f56e201fR1209) or not empty `FiniteSet`)  to check whether `args`  is modified or not.(means something reduced or not). If `args` is not modified(`new_args`  remain `false`) then `while` loop Line 1207 ends.

The issue will come when we do something like this : 

```
        while(new_args):
            for s in args:
                new_args = False
                args_without_s = set(args) - set(FiniteSet(s))
                args_without_s = FiniteSet(*[ar for ar in args_without_s])
                for t in args_without_s:
                    new_set = s._union(t)
                    # This returns None if s does not know how to intersect
                    # with t. Returns the newly intersected set otherwise
                    if new_set is not None:
                        if not isinstance(new_set, set):
                            new_set = set((new_set, ))
                        remv = set((s, t))
                        new_args = (set(args) - remv).union(new_set)
                        # this new_args should be used next time. `new_args` length is `len(args) - 1`. so this loop break.
                        # and goes to next if condition `if new_args:`.
                        break
                if new_args:
                    new_args = FiniteSet(*[ar for ar in new_args])
                    args = new_args
                    # if here is no break here, it will continue with the loop `for s in args` in Line 1208
                    # with this modified args. Which will return wrong result.

```

czgdp1807 commented:
Closing in favour of https://github.com/sympy/sympy/pull/18489
